src/boot/efi/measure.c

```diff
@@ -141,43 +141,79 @@ bool tpm_present(void) {
         return tcg2_interface_check() || tcg1_interface_check();
 }
 
-EFI_STATUS tpm_log_event(uint32_t pcrindex, EFI_PHYSICAL_ADDRESS buffer, UINTN buffer_size, const char16_t *description) {
-        EFI_TCG *tpm1;
+EFI_STATUS tpm_log_event(uint32_t pcrindex, EFI_PHYSICAL_ADDRESS buffer, UINTN buffer_size, const char16_t *description, bool *ret_measured) {
         EFI_TCG2 *tpm2;
+        EFI_STATUS ret;
 
         assert(description);
 
-        /* PCR disabled */
-        if (pcrindex == UINT32_MAX)
+        /* If EFI_SUCCESS is returned, will initialize ret_measured to true if we actually measured
```

Comment:
makes calling code simpler. i.e. the idea is that pcrindex == UINT32_MAX magically disables the measurement without this need special handling in all callers...

i.e. conceptually if we want something not to be treated as an error, we shouldn't return it as an error. because that forces calles to manually handle the errors manually even for clean codepaths, which is doable, but not ideal I think.

i.e. I am generally of the opinion that it's better to hide conditionalizations inside the functions to be conditionalized, instead of doing that in the caller. it's a scheme of "cleaner abstractions", as it allows you to see the general codeflow easier in the calling process, and not get lost in details about specific exceptional conditions...

Does that make any sense?

Reply:
I maintain that this makes all the code being touched ugly. Doing a `IN_SET(measure(...), EFI_SUCCESS, EFI_NOT_FOUND)` is imho much nicer to read than the `bool` dance you're doing and works similar to how `open()` works in userspace.

You also don't have to use `EFI_NOT_FOUND`. You can define a custom warning code and just use that. `EFI_STATUS` explicitly makes anything that doesn't have the MSB set a warning, which fits perfectly for this. In that case using `EFI_ERROR` to only filter for errors would even makes sense and keep the consumer side clean.

You should at least make the ret bool param mandatory where it makes sense (`pack_cpio` for example) to keep things simpler.

Reply:
Also note that there's stuff such as pack_cpio() that will set a `ret_measured`. Here in particular it becomes clear why the fact whether we actually measured something should not be propagated as an error code — it is not an *error* after all — but as a side-parameter. After all, the cpio packing should not abort just because the PCR measurement doesn't work.

Thats conceptually at the core of the problem: failure generally implies abortion of the regular execution paths of functions. And bypassing the PCR measurements should not do that. It should have almost no effect..